packages/tre/src/proxy.ts

```diff
@@ -0,0 +1,207 @@
+      await mkdir(path.dirname(cfPath), { recursive: true });
+      await writeFile(cfPath, cfText, "utf8");
+      console.log(grey("Updated `Request.cf` object cache!"));
+    } catch (e: any) {
+      console.log(
+        bold(
+          red(`Unable to fetch the \`Request.cf\` object! Falling back to a default placeholder...
+${dim(e.cause ? e.cause.stack : e.stack)}`)
+        )
+      );
+    }
+  }
+  // Initialise a proxy server that adds the `CF-Blob` header to runtime requests
+  createServer() {
+    const proxyWebSocketServer = new WebSocketServer({ noServer: true });
+    proxyWebSocketServer.on("connection", (client, request) => {
+      // Filter out browser WebSocket headers, since `ws` injects some. Leaving browser ones in causes key mismatches, especially with `Sec-WebSocket-Accept`
+      delete request.headers["sec-websocket-version"];
+      delete request.headers["sec-websocket-key"];
+      delete request.headers["sec-websocket-extensions"];
+      request.headers[CfHeader.Blob] = JSON.stringify(this.#cf);
+
+      const wsRuntimeEntryURL = new URL(this.runtimeURL!.href);
+      wsRuntimeEntryURL.protocol = "ws";
+
+      const runtime = new WebSocket(new URL(request.url!, wsRuntimeEntryURL), {
+        headers: request.headers,
+      });
+
+      // Proxy messages to and from the runtime
+      client.addEventListener("message", (e: MessageEvent) =>
+        runtime.send(e.data)
+      );
+      client.addEventListener("close", (e) => {
+        if (e.code === 1005 /* No Status Received */) {
+          runtime.close();
+        } else {
+          runtime.close(e.code, e.reason);
+        }
+      });
+
+      runtime.addEventListener("message", (e: MessageEvent) => {
+        client.send(e.data);
+      });
+      runtime.addEventListener("close", (e) => {
+        if (e.code === 1005 /* No Status Received */) {
+          client.close();
+        } else {
+          client.close(e.code, e.reason);
+        }
+      });
+    });
+    const server = http.createServer((originalRequest, originalResponse) => {
+      originalRequest.headers[CfHeader.Blob] = JSON.stringify(this.#cf);
+
+      const runtimeRequest = http.request(
+        {
+          hostname: this.runtimeURL!.hostname,
+          port: this.runtimeURL!.port,
+          path: originalRequest.url,
+          method: originalRequest.method,
+          headers: originalRequest.headers,
+        },
+        (runtimeResponse) => {
+          originalResponse.writeHead(
+            runtimeResponse.statusCode as number,
+            runtimeResponse.headers
+          );
+          runtimeResponse.pipe(originalResponse);
+        }
+      );
+
+      originalRequest.pipe(runtimeRequest);
+    });
+    // Handle websocket requests
+    server.on("upgrade", (request, socket, head) => {
+      proxyWebSocketServer.handleUpgrade(request, socket, head, (ws) => {
+        proxyWebSocketServer.emit("connection", ws, request);
+      });
+    });
```

Comment:
Ah, that's a good point—proxying less is definitely a good idea. I had been testing with a browser websocket, which doesn't expose much in the way of complexity (no access to headers etc...), so I hadn't run across these edge cases, but I suppose we should support arbitrary clients rather than just browser ones.

I can definitely switch over to using the `http-proxy` approach, but alternatively (and to get around any further edge cases that we might run across), what about stepping down the protocol stack slightly, and proxying the TCP socket instead? All we really need to change about the request is the `cf-blob` header, and we might be better off injecting that directly rather than going through Node's http stack?

```ts
function injectCfHeader(cf: IncomingRequestCfProperties) {
      return async function* (source: AsyncIterable<Buffer>) {
        for await (const chunk of source) {
          if (chunk.includes("\r\n\r\n")) {
            const index = chunk.indexOf("\r\n\r\n");
            const pre = Buffer.from(chunk.slice(0, index));
            const post = Buffer.from(chunk.slice(index));
            const header = Buffer.from(
              `\r\n${CfHeader.Blob}: ${JSON.stringify(cf)}`
            );
            yield Buffer.concat([pre, header, post]);
          } else {
            yield chunk;
          }
        }
      };
    }
    const tcpServer = net.createServer((conn) => {
      const runtime = net.createConnection(
        Number(this.runtimeURL!.port),
        this.runtimeURL!.hostname
      );
      conn.pipe(compose(injectCfHeader(this.#cf))).pipe(runtime);
      conn.on("end", () => runtime.end());
      conn.on("error", () => runtime.end());
      runtime.on("error", () => conn.end());
      runtime.on("end", () => conn.end());
      runtime.pipe(conn);
    });
```